sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodOverloadOptionalParameter.cs

```diff
@@ -148,15 +148,31 @@ private static bool IsMethodHidingOriginal(IMethodSymbol candidateHidingMethod,
         {
             return candidateHidingMethod.Parameters
                     .Zip(method.Parameters, (param1, param2) => new { param1, param2 })
-                    .All(p => AreParameterTypesEqual(p.param1, p.param2) &&
+                    .All(p => AreTypesEqual(p.param1.Type, p.param2.Type) &&
                               p.param1.IsOptional == p.param2.IsOptional);
         }
 
-        private static bool AreParameterTypesEqual(IParameterSymbol p1, IParameterSymbol p2)
+        private static bool AreTypesEqual(ITypeSymbol t1, ITypeSymbol t2)
         {
-            return Equals(p1.Type, p2.Type) ||
-                   p1.Type.Is(TypeKind.TypeParameter) && p2.Type.Is(TypeKind.TypeParameter) ||
-                   Equals(p1.Type.OriginalDefinition, p2.Type.OriginalDefinition);
+            return Equals(t1, t2) ||
+                   t1.Is(TypeKind.TypeParameter) && t2.Is(TypeKind.TypeParameter) ||
+                   AreGenericInstancesTypesEqual(t1, t2);
+        }
+
+        private static bool AreGenericInstancesTypesEqual(ITypeSymbol t1, ITypeSymbol t2)
+        {
+            if (t1.OriginalDefinition != t2.OriginalDefinition)
+            {
+                return false;
+            }
+            var named1 = t1 as INamedTypeSymbol;

```

Comment:
I would write the following as
```csharp
if (t1 is INamedTypeSymbol named1 &&
    t2 is INamedTypeSymbol named2)
{
    return named1.TypeArguments.SequenceEqual(named2.TypeArguments,
       (arg1, arg2) => AreTypesEqual(arg1, arg2));
}

return false;
```

Reply:
I know this is merged now, but what about types with more than 2 generic parameters?

Reply:
Hi @pms1969,

I think you might have misread the code. Indeed we are actually covering n parameters. If you look the code we actually group (zip) parameters from the method with the potential matching method and then for each parameter we check whether the `TypeArguments` are all equals (`named1.TypeArguments.SequenceEqual(named2.TypeArguments`).

Assuming we have the following 2 methods:
```csharp
void M1(Tuple<int, string, double> t, string msg = null) {}
void M1(Tuple<int, string, double> t) {}
```
We will do a grouping like this:
```csharp
(Tuple<int, string, double>, Tuple<int, string, double>), (string, nothing)
```
As the first group has generic parameters, we compare them all using `SequenceEquals`.

Hope this makes sense now?